plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,249 @@
+ */
+package io.trino.plugin.exchange;
+
+import io.airlift.log.Logger;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import io.trino.spi.exchange.ExchangeSink;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.crypto.SecretKey;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+public class FileSystemExchangeSink
+        implements ExchangeSink
+{
+    private static final Logger log = Logger.get(FileSystemExchangeSink.class);
+
+    public static final String COMMITTED_MARKER_FILE_NAME = "committed";
+    public static final String DATA_FILE_SUFFIX = ".data";
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FileSystemExchangeSink.class).instanceSize();
+
+    private final FileSystemExchangeStorage exchangeStorage;
+    private final URI outputDirectory;
+    private final int outputPartitionCount;
+    private final int numBuffers;
+    private final Optional<SecretKey> secretKey;
+    private final int writeBufferSize;
+
+    private final Map<Integer, ExchangeStorageWriter> writers = new ConcurrentHashMap<>();
+    private final Map<Integer, SliceOutput> pendingBuffers = new ConcurrentHashMap<>();
+    private final LinkedBlockingQueue<SliceOutput> freeBuffers = new LinkedBlockingQueue<>();
+    private volatile boolean committed;
+    private volatile boolean closed;
+    @GuardedBy("this")
+    private CompletableFuture<Void> blockedFuture = new CompletableFuture<>();
+
+    public FileSystemExchangeSink(FileSystemExchangeStorage exchangeStorage, URI outputDirectory, int outputPartitionCount, Optional<SecretKey> secretKey)
+    {
+        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
+        this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+        this.outputPartitionCount = outputPartitionCount;
+        this.numBuffers = outputPartitionCount * 2;
+        this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        this.writeBufferSize = exchangeStorage.getWriteBufferSizeInBytes();
+
+        for (int i = 0; i < numBuffers; ++i) {
+            freeBuffers.add(Slices.allocate(writeBufferSize).getOutput());
+        }
+    }
+
+    @Override
+    public CompletableFuture<?> isBlocked()
+    {
+        if (freeBuffers.isEmpty() && pendingBuffers.size() < outputPartitionCount) {
```

Comment:
This implementation isn't thread safe. Have you considered implementing it in a slightly different way:

1. At the beginning the `blockedFuture` is asigned to an unblocked future
2. When the freeBuffers pool becomes empty - a new `blockedFuture` is created in a blocked state
3. When the freeBuffers pool becomes not empty - an existing `blockedFuture` is unblocked
4. `isBlocked` method simply returns an existing future.

Reply:
You are right that the implementation is not thread safe, but it doesn't affect correctness as it returns NOT_BLOCKED when actually it's blocked in certain conditions. Current implementation is much like a best-effort hint.

The suggested approach works too but looks like we need to put all these operations under the same lock, right? I will test to see which one is better.

Reply:
I don't think we are able to use a global lock to achieve thread safety here with the use of a blocking queue. I'm sticking to best effort for now. This part will need more discussion.